sql/core/src/main/scala/org/apache/spark/sql/execution/CollectMetricsExec.scala

```diff
@@ -69,7 +69,7 @@ case class CollectMetricsExec(
       // - Performance issues due to excessive serialization.
```

Comment:
Reading the comments above, our goal is to only send back the accumulator updates to the driver side when the task completes, and the approach we took is to create a temp task local accumulator and set it to the real accumulator when the task ends.

I think there are three options:
1. allow merging accumulators in the executor side. But the consequence is not very clear.
2. find a way to share this temp task local accumulator between different partitions that run in the same task.
3. improve the accumulator framework to support "no update via heartbeat" feature natively, so that we don't need the temp task local accumulator.

Reply:
Maybe we can try option 1 first, as it's the simplest. We just need to update the `withSQLConf` function a little bit
```
  private[this] def withSQLConf[T](body: => T): T = {
    if (conf != null) {
      SQLConf.withExistingConf(conf)(body)
    } else {
      body
    }
  }
```

Reply:
About No. 2 above, `TaskContext` is apparently task local and can be shared among partitions even if a task can handle multiple partitions. So, I wonder a simple way is to have a general purpose task local storage in `TaskContext` and use it for task local accumulator.

No. 3 is too expensive to solve this problem isn't it?

Reply:
> Maybe we can try option 1 first, as it's the simplest. We just need to update the withSQLConf function a little bit

Hmm,  it seems the simplest way if we can.

Reply:
@cloud-fan I found we can't call `merge` from executor side. In this function, `mergeProjection` is referred and it is defined as follows.
https://github.com/apache/spark/blob/33f26275f4d65f54e68f38ba0d795396a5a4d2f4/sql/core/src/main/scala/org/apache/spark/sql/execution/AggregatingAccumulator.scala#L87-L89

In this lazy val, `mergeExpressions` is referred but it's null in executor side.
https://github.com/apache/spark/blob/33f26275f4d65f54e68f38ba0d795396a5a4d2f4/sql/core/src/main/scala/org/apache/spark/sql/execution/AggregatingAccumulator.scala#L36-L37

So, if we call `merge` from executor side, NPE is thrown.

Reply:
Let's remove `@transient`?

Reply:
> So, I wonder a simple way is to have a general purpose task local storage in TaskContext and use it for task local accumulator.

It needs a bit API design for the task local storage.

Reply:
> Let's remove @transient?

Is it acceptable? Anyway I'll try it.

Reply:
It is acceptable, just tried to trim down the accumulator as much as possible.